### PR TITLE
Oauth2 pkce faq

### LINKED ISSUE #2458: Document legacy/insecure resource settings
**Is your feature request related to a problem? Please describe.**

As a kanidm administrator, it feels "scary" to enable legacy or insecure options. Unfortunately, many oauth2 clients do not support the kanidm defaults. Further, when communicating with authors of oauth2 client applications, it would be helpful to have information we can provide to justify why we're asking for specific features.

**Describe the solution you'd like**

Please add to the FAQ explanations of the legacy and insecure options (disabling PKCE/crypto) , providing justification on why kanidm chooses these defaults. I would like this to target two audiences, kanidm administrators and oauth2 client developers. 

**Describe alternatives you've considered**

Status quo, no changes.

**Additional context**

### PR DESCRIPTION
**Pull request by Firstyear**
Fixes #2458 - add oauth2 pkce section to FAQ

Checklist

- [ x ] This pr contains no AI generated code
- [ x ] cargo fmt has been run
- [ ] cargo clippy has been run
- [ ] cargo test has been run and passes
- [ x ] book chapter included (if relevant)
- [ ] design document included (if relevant)

### CODE REVIEW

**Comment by Firstyear** [comment]
@adamcstephens

File: book/src/frequently_asked_questions.md [hunk 1]
```diff
@@ -39,6 +28,55 @@ Simply put, we are trying to use settings like `secure_cookies` to add constrain
...
+times.
+
+## OAuth2
+
+[RFC6819 - OAuth2 Threat Model and Security Considerations](https://www.rfc-editor.org/rfc/rfc6819)
+is a comprehensive and valuable resource discussing the security of OAuth2 and influences OpenID
+Connect as well. In general Kanidm follows and implements many of the recommendations in this
+document, as well as choosing not to implement certain known insecure OAuth2 features.
+
+### Why is disabling PKCE considered insecure?
```
Comment (adamcstephens) [inline book/src/frequently_asked_questions.md:47]: Should this section link to https://kanidm.github.io/kanidm/master/integrations/oauth2.html#extended-options-for-legacy-clients ?

**Review by adamcstephens — COMMENTED** [review]
I was also thinking about the legacy crypto setting, `warning-enable-legacy-crypto`. Would it make sense to make an explicit note about the crypto levels? I've had at least one dev tell me it wasn't a big deal to use what is labeled as legacy by kanidm.

File: book/src/frequently_asked_questions.md (hunk 1 above)
Comment (Firstyear) [reply book/src/frequently_asked_questions.md:47]: done

**Comment by Firstyear** [comment]
> I was also thinking about the legacy crypto setting, `warning-enable-legacy-crypto`. Would it make sense to make an explicit note about the crypto levels? I've had at least one dev tell me it wasn't a big deal to use what is labeled as legacy by kanidm.

Doing this now :)

**Review by adamcstephens — APPROVED** [review]
Looks good to me.

File: book/src/frequently_asked_questions.md [hunk 2]
```diff
@@ -39,6 +28,81 @@ Simply put, we are trying to use settings like `secure_cookies` to add constrain
...
+For confidential clients (refered to as a `basic` client in Kanidm due to the use of HTTP Basic for
+`client_id+secret` presentation) PKCE may optionally be disabled. This can allow authorisation code
+attacks to be carried out - however _if_ TLS is used and the `client_secret` never leaks, then these
+attacks will not be possible. Since there are many public references to system administrators
+mishandling secrets such as these so we should not rely on this as our sole defence.
+
+For public clients (which have no `client_id` authentication) we strictly enforce PKCE since
+disclosure of the authorisation code to an attacker will allow them to perform the code exchange.
+
+OpenID connect internally has a `nonce` parameter in it's operations. Commonly it is argued that
```
Comment (yaleman) [inline book/src/frequently_asked_questions.md:71]:
```suggestion
OpenID connect internally has a `nonce` parameter in its operations. Commonly it is argued that
```

File: book/src/frequently_asked_questions.md [hunk 3]
```diff
@@ -39,6 +28,81 @@ Simply put, we are trying to use settings like `secure_cookies` to add constrain
...
+
+We would strongly encourage OAuth2 client implementations to implement and support PKCE, as it
+provides defense in depth to known and exploited authorisation code interception attacks.
+
+### Why is RSA considered legacy
+
+While RSA is cryptographically sound, to achieve the same level as security as ECDSA it requires
+signatures and keys that are significantly larger. This has costs for network transmission and CPU
+time to verify these signatures. At this time (2024) to achieve the same level of security as a 256
+bit ECDSA, RSA requires a 3072 bit key. Similar a 384 bit ECDSA key requires a 8192 bit RSA for
```
Comment (yaleman) [inline book/src/frequently_asked_questions.md:85]:
```suggestion
bit ECDSA, RSA requires a 3072 bit key. Similarly, a 384 bit ECDSA key requires a 8192 bit RSA for
```

File: book/src/frequently_asked_questions.md [hunk 4]
```diff
@@ -39,6 +28,81 @@ Simply put, we are trying to use settings like `secure_cookies` to add constrain
...
+equivalent cryptographic strength, and a 521 bit ECDSA key would likely require a 16884 bit RSA key
+(or greater).
+
+This means that going forward more applications will require ECDSA over RSA due to it's increased
```
Comment (yaleman) [inline book/src/frequently_asked_questions.md:89]:
```suggestion
This means that going forward more applications will require ECDSA over RSA due to its increased
```

File: book/src/frequently_asked_questions.md [hunk 5]
```diff
@@ -39,6 +28,81 @@ Simply put, we are trying to use settings like `secure_cookies` to add constrain
...
+strength for significantly faster and smaller key sizes.
+
+Where this has more serious costs is our future desire to add support for Hardware Security Modules.
+Since RSA keys are much larger on these devices it may significantly impact performance of the HSM
+and may also limit the amount of keys we can store on the device. In the case of some HSM models,
+they do not even support RSA keys up to 8192 bits (but they do support ECDSA 384 and 521). An
+example of this is TPMs, which only support up to 4096 bit RSA keys at this time.
+
+As a result, we want to "guide" people toward smaller, faster and more secure cryptographic
```
Comment (yaleman) [inline book/src/frequently_asked_questions.md:98]:
```suggestion
As a result, we want to guide people toward smaller, faster and more secure cryptographic
```

**Review by yaleman — CHANGES_REQUESTED** [review]

**Review by yaleman — APPROVED** [review]

**Pull request merged**